plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/GcpLocalRunTab.java

```diff
@@ -0,0 +1,304 @@
+import org.eclipse.swt.widgets.Label;
+import org.eclipse.swt.widgets.Text;
+import org.eclipse.ui.PlatformUI;
+
+public class GcpLocalRunTab extends AbstractLaunchConfigurationTab {
+
+  private static final Logger logger = Logger.getLogger(GcpLocalRunTab.class.getName());
+
+  private static final String ATTRIBUTE_ACCOUNT_EMAIL =
+      "com.google.cloud.tools.eclipse.gcpEmulation.accountEmail"; //$NON-NLS-1$
+
+  private static final String PROJECT_ID_ENVIRONMENT_VARIABLE =
+      "GOOGLE_CLOUD_PROJECT"; //$NON-NLS-1$
+  private static final String SERVICE_KEY_ENVIRONMENT_VARIABLE =
+      "GOOGLE_APPLICATION_CREDENTIALS"; //$NON-NLS-1$
+
+  private final EnvironmentTab environmentTab;
+  private final IGoogleLoginService loginService;
+  private final ProjectRepository projectRepository;
+
+  private AccountSelector accountSelector;
+  private ProjectSelector projectSelector;
+  private Text serviceKeyInput;
+
+  private Image gcpIcon;
+
+  // We set up intermediary models between a run configuration and UI components for certain values,
+  // because, e.g., the account selector cannot load an email if it is not logged in. In such a
+  // case, although nothing is selected in the account selector, we should not clear the email saved
+  // in the run configuration.
+  private String accountEmailModel;
+  private String gcpProjectIdModel;
+  // To prevent updating above models when programmatically setting up UI components.
+  private boolean initializingUiValues;
+
+  private boolean activated;  // to avoid https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/2568#discussion_r150128582
+
+  public GcpLocalRunTab(EnvironmentTab environmentTab) {
+    this(environmentTab, PlatformUI.getWorkbench().getService(IGoogleLoginService.class),
+        new ProjectRepository(PlatformUI.getWorkbench().getService(IGoogleApiFactory.class)));
+  }
+
+  @VisibleForTesting
+  GcpLocalRunTab(EnvironmentTab environmentTab,
+      IGoogleLoginService loginService, ProjectRepository projectRepository) {
+    this.environmentTab = environmentTab;
+    this.loginService = loginService;
+    this.projectRepository = projectRepository;
+  }
+
+  @Override
+  public void createControl(Composite parent) {
+    Composite composite = new Composite(parent, SWT.NONE);
+
+    // Account row
+    new Label(composite, SWT.LEAD).setText(Messages.getString("label.account")); //$NON-NLS-1$
+    accountSelector = new AccountSelector(composite, loginService);
+    accountSelector.addSelectionListener(new Runnable() {
+      @Override
+      public void run() {
+        updateProjectSelector();
+
+        if (!initializingUiValues) {
+          boolean somethingSelected = !accountSelector.getSelectedEmail().isEmpty();
+          boolean savedEmailAvailable = accountSelector.isEmailAvailable(accountEmailModel);
+          if (somethingSelected || savedEmailAvailable) {
+            accountEmailModel = accountSelector.getSelectedEmail();
```

Comment:
This doesn't seem right: if no account is currently selected, but our saved-email (`accountEmailModel`) is available, then we throw away our saved-email.

Ohhh! I see: `savedEmailAvailable==true` means the selector has been populated, and so we now update the `accountEmailModel` to keep it in sync with changes.  A comment would help here, and below for the project selector.

Reply:
Right. This covers the case where users explicitly de-selected any selection. In this case, we toss the saved value. (Account deselection is not implemented yet, but we do have an issue for this enhancement). I'll add a comment.